demo-circuit/src/naive_threshold_sig_w_key_rotation/tests/mod.rs

```diff
@@ -0,0 +1,1340 @@
+mod first_cert;
+mod utils;
+
+use algebra::{Field, ToConstraintField, UniformRand};
+use cctp_primitives::{
+    proving_system::{
+        compute_proof_vk_size,
+        init::{get_g1_committer_key, load_g1_committer_key},
+        ProvingSystem,
+    },
+    utils::commitment_tree::hash_vec,
+};
+use primitives::schnorr::field_based_schnorr::FieldBasedSchnorrPk;
+use primitives::{
+    schnorr::field_based_schnorr::FieldBasedSchnorrSignature, FieldBasedHash,
+    FieldBasedSignatureScheme,
+};
+use r1cs_core::debug_circuit;
+use rand::thread_rng;
+use serial_test::serial;
+
+use crate::{
+    common::{WithdrawalCertificateData, NULL_CONST},
+    create_msg_to_sign,
+    type_mapping::*,
+};
+use rand::Rng;
+
+use crate::{MAX_SEGMENT_SIZE, SUPPORTED_SEGMENT_SIZE};
+
+use super::{data_structures::ValidatorKeysUpdates, NaiveThresholdSignatureWKeyRotation};
+
+pub const VALIDATORS_SIZE: usize = 6;
+
+#[derive(Copy, Clone, Debug)]
+enum TestDataType {
+    AllRandomFirstCert,
+    AllRandomNextCert,
+    ValidSigEqThresholdNextCert,
+    ValidSigGreaterThresholdNextCert,
+    ValidSigSmallerThresholdNextCert,
+    ValidSigAndThresholdEqZeroNextCert,
+    ValidSigAndThresholdEqValidatorsNextCert,
+    ValidSigEqValidatorsThresholdEqZeroNextCert,
+    ValidSigEqZeroThresholdEqValidatorsNextCert,
+    ChangeMasterKeysNextCert,
+    ChangeSigningAndMasterKeysNextCert,
```

Comment:
I would add also a couple of test cases to update the signing/master keys with the first certificate, which I think should be allowed by the specification